### PR TITLE
`pow`

### PR DESCRIPTION
**Pull request by jeksterslab**
I've started working on `pow.default` and `pow.yac_symbol` for `MatrixPower`. If this is a worthwhile method I'll go ahead and write some tests.

### CODE REVIEW

**Comment by mikldk** [comment]
Thanks, @jeksterslab . This is interesting, too. 

Your current implementation of `pow.default()` can be made more efficient, read e.g. <https://en.wikipedia.org/wiki/Exponentiation_by_squaring>. There are multiple ways of implementing more efficient schemes; one of the basic ideas is that e.g. `x^9 = x^4 * x^4 * x` so first calculate `x^2`, then `x^4 = x^2 * x^2`. The current method can e.g. be moved to `tests` and used to compare a more efficient one with.

**Comment by jeksterslab** [comment]
I have added preliminary tests for `pow`. I still need to see why `pow.yac_symbol` freezes my machine when I use negative exponents.

**Comment by jeksterslab** [comment]
Raising `x` to a negative power `n` using `y_fn` or `yac_str` freezes my machine. I double check `yacas` and everything is in order on that end. My hunch is that there is an issue regarding how the argument `n` is interpreted by `y_fn` or `yac_str` when it is a negative number.

```
library(Ryacas)

A <- mtcars[, c(1, 3, 4, 5, 6, 7)]
x <- cov(A)

# Positive Power - Works like a charm
n <- 2
y_fn(ysym(x), "MatrixPower", n)
yac_str(paste0("MatrixPower(", ysym(x), ",", n, ")"))

# Negative Power - Freezes my machine
n <- -1
y_fn(ysym(x), "MatrixPower", n)
yac_str(paste0("MatrixPower(", ysym(x), ",", n, ")"))
```

**Comment by mikldk** [comment]
@jeksterslab Not sure about the hanging. Can you try again after `yacas` update to 1.9.2 dev. in d1c2d7d1c223240e9b7f268f382009b057836b38?

**Comment by jeksterslab** [comment]
I'm still having issues with negative powers even with the update. I'll investigate the issue further.

**Comment by jeksterslab** [comment]
@mikldk can you try to run the code below on your machine? I can't seem to isolate the problem with negative exponents on my end.

```
library(Ryacas)

A <- mtcars[, c(1, 3, 4, 5, 6, 7)]
x <- cov(A)

# Positive Power - Works like a charm
n <- 2
y_fn(ysym(x), "MatrixPower", n)
yac_str(paste0("MatrixPower(", ysym(x), ",", n, ")"))

# Negative Power - Freezes my machine
n <- -1
y_fn(ysym(x), "MatrixPower", n)
yac_str(paste0("MatrixPower(", ysym(x), ",", n, ")"))
```

**Comment by jeksterslab** [comment]
I tested `pow.yac_symbol` with a 2 by 2 matrix using a negative `n` power and it worked properly.

Apparently the inverse of `x` in the case above takes forever on my machine using any of the following: `y_fn(ysym(x), "MatrixPower", -1)`, `yac_str(paste0("MatrixPower(", ysym(x), ",", -1, ")"))`, and `solve(x)`

I also tried the same matrix as input in `yacas` (see below) and the same happens. There is nothing wrong with the implementation of `pow.yac_symbol`. It just takes to long to obtain a solution in `yacas` for this case. Obtaining the inverse of a matrix is not a trivial task.

```
x := {{36.32,-633.09,-320.73,2.19,-5.11,4.50},
        {-633.09, 15360.79, 6721.15, -47.06, 107.68, -96.05},
        {-320.73, 6721.15, 4700.86, -16.45, 44.19, -86.77},
        {2.19, -47.06, -16.45, 0.28, -0.37, 0.08},
        {-5.11, 107.68, 44.19, -0.37, 0.95, -0.30},
        {4.50, -96.05, -86.77, 0.08, -0.30, 3.19}};
Inverse(x);
MatrixPower(x, -1);
```

Note, however, that `pow.default` does not have this issue.

In summary,

- `pow.yac_symbol` is in order, however, it might have performance issues with negative exponents for larger matrix inputs and
- `pow.default` using exponentiation by squaring is fast enough for most applications.

Let me know which direction you wish to take.

**Comment by mikldk** [comment]
@grzegorzmazur do you have any input on this?

**Comment by grzegorzmazur** [comment]
The matrix inverse in yacas is fantastic, thanks to great contribution by @yijunyu but in the end it uses yacas built-in matrix multiplication. And this is unfortunately not fantastic at all, at least efficiency-wise. I think I can improve things somewhat in next couple of days, stay tuned.

**Comment by jeksterslab** [comment]
You guys are doing an awesome job with `yacas`!!! As usual, I can help test any changes in `yacas` via `Ryacas`.

**Comment by grzegorzmazur** [comment]
OK, I've mitigated the matrix inverse efficiency problem in https://github.com/grzegorzmazur/yacas/commit/71f6d08ebcc997c248ff0e3f8b47b6b896596daf It's still far from good, but on the example above the speedup is 20-fold, which should be enough to make it usable.

**Comment by grzegorzmazur** [comment]
and one more improvement, actually a change in semantics which was long overdue, with a nice side effect of improving efficiency for the example above by another order of magnitude (and giving results in a nicer form): https://github.com/grzegorzmazur/yacas/commit/ac280d0a16bbe81ec86870325403a65aee6e30fd

**Comment by jeksterslab** [comment]
Thanks @grzegorzmazur. I'll do some benchmarks in `R` once this fix is made available in `Ryacas`.

**Comment by mikldk** [comment]
Thanks, @grzegorzmazur . Now updated, @jeksterslab , in 37fa4f5c6716af53a71cb4703ea09cae8b892c8a.

**Comment by jeksterslab** [comment]
There's a huge improvement in performance since the update. Below are some benchmarks with positive and negative powers.

# Benchmark 1 - Positive Power

```
library(Ryacas)
library(microbenchmark)

A <- mtcars[, c(1, 3, 4, 5, 6, 7)]
x <- cov(A)

n <- 5
microbenchmark(
  pow(x, n),
  pow(ysym(x), n),
  times = 1000
)

Unit: microseconds
            expr       min        lq       mean    median         uq       max neval
       pow(x, n)    33.183    38.821    57.2089    61.998    67.8655  2382.409 1000
 pow(ysym(x), n) 25955.561 26365.360 26511.3803 26434.906 26531.3090 33788.084 1000
```

# Benchmark 2 - Negative Power

```
n <- -5
microbenchmark(
  pow(x, n),
  pow(ysym(x), n),
  times = 1000
)

Unit: microseconds
            expr       min        lq       mean    median         uq       max neval
       pow(x, n)    57.983    70.402   111.6013   122.921   151.2615   2263.35 1000
 pow(ysym(x), n) 62528.944 64680.171 66505.0019 66611.215 67760.9395 105248.55 1000
``` 

# Session Info  
  
```
R version 4.0.4 (2021-02-15)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Arch Linux

Matrix products: default
BLAS:   /usr/lib/libopenblasp-r0.3.13.so
LAPACK: /usr/lib/liblapack.so.3.9.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=C              
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] colorout_1.2-2          jeksterslabRutils_0.9.0

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.6        knitr_1.30        magrittr_2.0.1    doParallel_1.0.16
 [5] xtable_1.8-4      R6_2.5.0          rlang_0.4.10      fastmap_1.0.1    
 [9] foreach_1.5.1     styler_1.3.2      tools_4.0.4       parallel_4.0.4   
[13] xfun_0.20         htmltools_0.5.1   iterators_1.0.13  ellipsis_0.3.1   
[17] readxl_1.3.1      digest_0.6.27     lifecycle_0.2.0   tibble_3.0.6     
[21] crayon_1.4.1      shiny_1.5.0       purrr_0.3.4       later_1.1.0.1    
[25] vctrs_0.3.6       promises_1.1.1    codetools_0.2-18  evaluate_0.14    
[29] mime_0.9          rmarkdown_2.6     compiler_4.0.4    pillar_1.4.7     
[33] cellranger_1.1.0  backports_1.2.1   doMC_1.3.7        httpuv_1.5.5     
[37] pkgconfig_2.0.3 
```

**Comment by jeksterslab** [comment]
All the tests and examples are good for `pow`, `vec`, and `vech`.

**Comment by grzegorzmazur** [comment]
OK, this isn't that bad. Obviously, yacas will never be able to compete speed-wise with blas, but at least the floating-point operations should be fast enough to be usable for simple calculations.

**Comment by jeksterslab** [comment]
@grzegorzmazur yup I agree with you blas is blas. I guess what I wanted to show is that `pow.yac_symbol` can hang in there with `pow.default`. Here is the same benchmark using `seconds` as the unit. In both cases, `pow.yac_symbol`  is within a  fraction of a second which is fast enough for most use-case.

```
library(Ryacas)
library(microbenchmark)

A <- mtcars[, c(1, 3, 4, 5, 6, 7)]
x <- cov(A)

n <- 5
microbenchmark(
  pow(x, n),
  pow(ysym(x), n),
  times = 1000,
  unit = "s"
)

Unit: seconds
            expr         min           lq         mean     median          uq
       pow(x, n) 0.000108825 0.0001297645 0.0001609462 0.00014781 0.000177036
 pow(ysym(x), n) 0.025678974 0.0261583515 0.0264407946 0.02625287 0.026345138
        max neval
 0.00241687  1000
 0.03992452  1000

n <- -5
microbenchmark(
  pow(x, n),
  pow(ysym(x), n),
  times = 1000,
  unit = "s"
)

Unit: seconds
            expr         min          lq         mean       median          uq
       pow(x, n) 0.000113435 0.000128947 0.0001822814 0.0001818915 0.000228381
 pow(ysym(x), n) 0.068446725 0.069834913 0.0717337738 0.0712982140 0.073342775
        max neval
 0.00237628  1000
 0.12809830  1000
```

**Comment by jeksterslab** [comment]
`pow`, `vec`, and `vech` looks good on my end. Let me know if there is anything else you need @mikldk so we can merge `pow`, `vec`, and `vech` with the main branch.

**Comment by grzegorzmazur** [comment]
@jeksterslab right, no offence intended, I just wanted to point out that yacas does not even pretend to play in the same league when it comes to floating point calculations on array, and that I'm actually quite happy with the performance as is. And thanks a lot for your work :)

**Comment by jeksterslab** [comment]
@grzegorzmazur no offense taken. Hopefully my questions are helping improve `yacas` and `Ryacas`.

**Comment by mikldk** [comment]
@jeksterslab Is this still good to be merged? Nothing in `yacas` that we need to pull in before merging?

**Comment by jeksterslab** [comment]
@mikldk I think we're good to go.
